### PR TITLE
Simpleinjector reimplementation

### PR DESCRIPTION
**Pull request by jposert**
**What kind of change does this PR introduce?**
Bugfix

**What is the current behavior?**
Currently there's no way to use SimpleInjector in projects.

**What is the new behavior?**
SimpleInjector works well with Splat & ReactiveUI initializations.

**What might this PR break?**
Probably nothing.

**Please check if the PR fulfills these requirements**
- [ x] Tests for the changes have been added (for bug fixes / features)
- [ x] Docs have been added / updated (for bug fixes / features)

**Other information**:
I've tested the solution with my very small project based on WPF - all works okay. Anyway, I've no clue about performance of the solution (probably most important thing on mobile platforms).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/reactiveui/splat?pullRequest=631) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/reactiveui/splat?pullRequest=631) before we can accept your contribution.<br/><br/>:x: jposert [sign now](https://cla.dotnetfoundation.org/reactiveui/splat?pullRequest=631)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/reactiveui/splat?pullRequest=631) it.</sub>

**Comment by glennawatson** [comment]
There's some failing unit tests unfortunately. You can click the build action to see the details. 

Thanks for the work so far

**Comment by jposert** [comment]
I made a little mistake and i pushed  invalid code (sorry). For now, everything should be okay but the build is failing - and I'm not sure why or how to fix it. It's all building on my machine and it passes all the tests successfully.

**Comment by glennawatson** [comment]
The latest is a build error that is on and off we've seen before with Nerdbank git versioning tool we use. I might make a report to the library author.

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    /// sad.
+    /// </summary>
+#pragma warning disable CA1063 // Implement IDisposable Correctly
+    public class SimpleInjectorInitializer : IDependencyResolver
+#pragma warning restore CA1063 // Implement IDisposable Correctly
+    {
+        /// <summary>
+        /// Gets dictionary of registered factories.
+        /// </summary>
+        public Dictionary<Type, List<Func<object>>> RegisteredFactories { get; }
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:23]: Can you take a look at the lock logic in https://github.com/reactiveui/splat/blob/7af765b6ac360c93dc217d6380cc045c23c2401a/src/Splat.Autofac/AutofacDependencyResolver.cs

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            = new Dictionary<Type, List<Func<object>>>();
+
+        /// <inheritdoc />
+        public object GetService(Type serviceType, string contract = null)
+        {
+            Func<object> fact = RegisteredFactories[serviceType].LastOrDefault();
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:29]: Thread safety tweak needed with lock see https://github.com/reactiveui/splat/blob/7af765b6ac360c93dc217d6380cc045c23c2401a/src/Splat.Autofac/AutofacDependencyResolver.cs#L55-L61 for example

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            return fact?.Invoke();
+        }
+
+        /// <inheritdoc/>
+        public IEnumerable<object> GetServices(Type serviceType, string contract = null)
+        {
+            return RegisteredFactories[serviceType]
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:36]: thread safety tweak needed
https://github.com/reactiveui/splat/blob/7af765b6ac360c93dc217d6380cc045c23c2401a/src/Splat.Autofac/AutofacDependencyResolver.cs#L88-L103

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                .Select(n => n());
+        }
+
+        /// <inheritdoc />
+        public bool HasRegistration(Type serviceType, string contract = null)
+        {
+            return RegisteredFactories.TryGetValue(serviceType, out List<Func<object>> values)
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:43]: thread safety tweak
https://github.com/reactiveui/splat/blob/7af765b6ac360c93dc217d6380cc045c23c2401a/src/Splat.Autofac/AutofacDependencyResolver.cs#L106-L116

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                   && values.Any();
+        }
+
+        /// <inheritdoc />
+        public void Register(Func<object> factory, Type serviceType, string contract = null)
+        {
+            if (!RegisteredFactories.ContainsKey(serviceType))
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:50]: thread safety tweak needed please :)

File: src/Splat.SimpleInjector/SimpleInjectorInitializer.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        /// <inheritdoc />
+        public void UnregisterCurrent(Type serviceType, string contract = null)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        public void UnregisterAll(Type serviceType, string contract = null)
+        {
+            if (RegisteredFactories.ContainsKey(serviceType))
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/SimpleInjectorInitializer.cs:67]: thread safety tweak needed please :)

File: src/Splat.SimpleInjector/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+### Register the Container
+
+Because of internal SimpleInjector behaviours in order to register it as a container intermediate "dummy" container has to be used - `SimpleInjectorInitializer`.
+
+```cs
+SimpleInjectorInitializer initializer = new SimpleInjectorInitializer();
+Locator.SetLocator(initializer);
+
+// Register ReactiveUI dependencies
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/README.md:15]: are these 2 initialize lines actually needed? Splat and ReactiveUI would do these during normal operation. Is there a nuance we need to document so people are aware?

File: src/Splat.SimpleInjector/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+// Actual SimpleInjector registration
+container.UseSimpleInjectorDependencyResolver(initializer);
+
+// SimpleInjector dependency registrations
+container.Register<Example>();
+container.Register<Example2>();
+
+// Optional overriding default implementations
+container.Options.AllowOverridingRegistrations = true;
```
Comment (dpvreony) [inline src/Splat.SimpleInjector/README.md:33]: i appreciate AllowOverridingRegistrations is "optional", but from what i've read of an issue, this actually makes sense to use in certain scenarios, could we just add a little bit of extra detail to the comment. helps for knowledge retention, dev journey and if we can revisit to overcome the issue in future :)

**Review by dpvreony — CHANGES_REQUESTED** [review]
thanks, few threading tweaks and some bits around the read me just to expand\ clarify knowledge

**Comment by dpvreony** [comment]
housekeeping comment
closes #597

**Comment by dpvreony** [comment]
cleaned up. documented weakness in dupe registrations

**Review by dpvreony — APPROVED** [review]

**Pull request merged**